jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -51,12 +55,78 @@ public Class findClass(@Nullable String name) {
     }
   }
 
+  /**
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} plugin provided by {@code projectProperties}.
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class is not valid, or throws an error if no valid main class is not found.
+   */
+  public static String resolveMainClass(
```

Comment:
I feel like it might make sense to separate these into two classes - one for searching for a main method class, and one for the actual logic of inferring a main class - these seem like two distinct things on different levels of abstraction.

Reply:
I'm not sure. `findMainClasses` is only ever called in `resolveMainClass` (and also in the tests, of course), so I see it more as an implementation detail rather than something that belongs on a separate level of abstraction. The functionality is different enough that yeah, maybe it makes sense to break it out by itself, but I'd rather avoid spreading it out when it's only used in that one place. We can split it out later if we need to (maybe when jib-core becomes a standalone library).